packages/www/pages/mint-nft/index.tsx

```diff
@@ -62,89 +60,55 @@ const networks = {
 type SupportedChainIDs = keyof typeof networks;
 type NetworkInfo = typeof networks[SupportedChainIDs];
 
-async function getMintedTokenIdOnce(
-  videoNft: Contract,
-  sender: string,
-  txReceipt: TransactionReceipt
-) {
-  const events = await videoNft.getPastEvents("Mint", {
-    filter: { sender },
-    fromBlock: txReceipt.blockNumber,
-    toBlock: txReceipt.blockNumber,
-  });
-  const event = events.find(
-    (ev) => ev.transactionHash === txReceipt.transactionHash
-  );
-  return event?.returnValues?.tokenId as string;
-}
-
-async function getMintedTokenId(
-  videoNft: Contract,
-  sender: string,
-  txReceipt: TransactionReceipt,
-  logger: (log: JSX.Element | string) => void
-) {
-  const maxAttempts = 5;
-  const retryDelayMs = 5 * 1000;
-  const sleep = (ms: number) => new Promise((res) => setTimeout(res, ms));
-
-  let lastErr: Error | undefined;
-  for (let a = 1; a <= maxAttempts; a++) {
-    if (a > 1) {
-      await sleep(retryDelayMs);
-    }
-    try {
-      const tokenId = await getMintedTokenIdOnce(videoNft, sender, txReceipt);
-      if (!tokenId) {
-        throw new Error("Missing tokenId");
-      }
-      return tokenId;
-    } catch (err) {
-      lastErr = err;
-    }
-  }
-  logger(`Error getting minted token ID: ${lastErr}`);
-  return null;
-}
-
 async function mintNft(
-  web3: Web3,
+  provider: ethers.providers.Web3Provider,
   contractAddress: string,
-  from: string,
   to: string,
   tokenUri: string,
   logger: (log: JSX.Element | string) => void,
   network: NetworkInfo
 ) {
   try {
     logger("Started mint transaction...");
-    const videoNft = new web3.eth.Contract(videoNftAbi as any, contractAddress);
-    const gasPrice = Math.round(1.2 * parseInt(await web3.eth.getGasPrice()));
-    const transaction = {
-      to: contractAddress,
-      gas: 500000,
-      maxPriorityFeePerGas: gasPrice,
-      data: videoNft.methods.mint(to, tokenUri).encodeABI(),
-    };
-    const nonce = await web3.eth.getTransactionCount(from, "latest");
-    const tx = {
-      ...transaction,
-      from,
-      nonce,
-    };
-    const receipt = await web3.eth.sendTransaction(tx);
+    const signer = provider.getSigner();
+    const videoNft = new ethers.Contract(
+      contractAddress,
+      videoNftAbi,
+      provider
+    ).connect(signer);
+
+    const selfAddr = await signer.getAddress();
+    let cancelTokenId: () => void;
+    const tokenIdPromise = new Promise<number | null>((resolve, reject) => {
+      const handler = (f, t, mintedUri: string, tokenId: ethers.BigNumber) => {
+        if (mintedUri !== tokenUri) return;
+        videoNft.off(filter, handler);
+        resolve(tokenId.toNumber());
+      };
+      cancelTokenId = () => {
+        videoNft.off(filter, handler);
+        resolve(null);
+      };
+      const filter = videoNft.filters.Mint(selfAddr, selfAddr, null, null);
+      videoNft.once(filter, handler);
+    });
```

Comment:
@adamsoffer so even though this is so much better than the polling of historic events, I still don't like this code. Seems very much like a hack on top of an API meant for streaming but not for getting a single event. 

I could not add a filter by `tokenUri` and just use `once` instead of `on`/`off` since the `tokenUri` field is un-indexable. I theoretically could make the field indexable in the smart contract but that would seem like a overfit into a bad interface or bad code here.

The cancel/timeout logic also seems pretty hacky, with the leak of the `resolve` function of the promise which we set below. 

All ideas welcome!

Reply:
Gotcha - yeah that field would have to be indexable. I'm cool with this hack for now if you are.

Reply:
Just found a much cleaner way to get the `Mint` event from the transaction receipt! ethers ❤️ 

Committed here: https://github.com/livepeer/livepeer-com/pull/976/commits/540ef4279e6030fb36c20ebad28966071713129d